File: core/shared/src/main/scala/diplomacy/variant/standard/rule/RetreatValidator.scala
package diplomacy.variant.standard.rule

import scala.collection.{ mutable => cmutable }

import diplomacy.rule.InvalidOrderMessage
import diplomacy.board.Power

class RetreatValidator[Turn_ <: Turn, Power_ <: Power] extends Validator[Turn_, Power_] {
  def unitsRequiringOrder(board: Board): Set[DiplomacyUnit] = (board.unitStatuses collect {
    case (unit, UnitStatus.Dislodged(_)) => unit
  })(collection.breakOut)

  def errorMessageOfOrder(board: Board)(order: Order): Option[InvalidOrderMessage] = {
    // The order is invalid if order.unit is not dislodged
    val dislodgedUnits = (board.unitStatuses collect {
      case (unit, x: UnitStatus.Dislodged[Power]) => unit -> x.attackedFrom 
    }).toMap
    if (!(dislodgedUnits contains order.unit)) {
      Option(InvalidOrderMessage(s"${order.unit} is not dislodged."))
    } else {
      order match {
        case Order.Retreat(unit, destination) =>
          if (
            !(StandardRuleUtils.locationsToRetreat(board)(unit, dislodgedUnits(unit)) contains destination)
          ) {
            Option(InvalidOrderMessage(s"${unit} cannot retreat to ${destination}."))
          } else {
            None
          }
        case Order.Disband(_) => None
        case _ => Option(InvalidOrderMessage(s"${order} is not order for Retreat phase."))
      }
    }
  }
  def errorMessageOfOrders(board: Board)(order: Set[Order]): Option[InvalidOrderMessage] = {
    board.unitStatuses collectFirst {
      case (unit, UnitStatus.Dislodged(_)) if order forall { _.unit != unit } =>
        InvalidOrderMessage(s"${unit} has no order.")
    }
  }
}

File: src/variant/standard/rule/retreat-validator.ts
import { Validator } from "./validator"
import { StandardRuleUtils } from "./standard-rule-utils"
import { OrderType, Order, Disband, Retreat } from "./order"
import { Dislodged } from "./dislodged"
import { Board, Unit } from "./types"

export class RetreatValidator<Power> implements Validator<Power> {
  unitsRequiringOrder (board: Board<Power>) {
    return new Set([...board.unitStatuses].map(elem => elem[0]))
  }
  errorMessageOfOrder (board: Board<Power>, order: Order<Power>) {
    // The order is invalid if order.unit is not dislodged
    const dislodged = [...board.unitStatuses].some(elem => {
      const unit = elem[0]
      return (unit.militaryBranch === order.unit.militaryBranch) &&
        (unit.location === order.unit.location) &&
        (unit.power === order.unit.power)
    })

    if (!dislodged) {
      return `${order.unit} is not dislodged`
    }

    // unitStatuses.find(...) is not null because disloged is true
    const status = (<[Unit<Power>, Dislodged<Power>]>[...board.unitStatuses].find(elem => {
      const unit = elem[0]
      return (unit.militaryBranch === order.unit.militaryBranch) &&
        (unit.location === order.unit.location) &&
        (unit.power === order.unit.power)
    }))[1]

    if (order instanceof Retreat) {
      const ls = StandardRuleUtils.locationsToRetreat(board, order.unit, status.attackedFrom)
      if (!ls.has(order.destination)) {
        return `${order.unit} cannot retreat to ${order.destination}`
      }
    } else if (!(order instanceof Disband)) {
      return `${order} is for Retreat phase`
    }
    return null
  }

  errorMessageOfOrders (board: Board<Power>, orders: Set<Order<Power>>) {
    for (let elem of [...board.unitStatuses]) {
      const [unit, status] = elem
      const hasOrder = [...orders].some(order => {
        return (unit.militaryBranch === order.unit.militaryBranch) &&
          (unit.location === order.unit.location) &&
          (unit.power === order.unit.power)
      })

      if (!hasOrder) {
        return `${unit} has no order`
      }
    }
    return null
  }
}
